### PR TITLE
Feature/dapps page

### PR DESCRIPTION
**Pull request by sirius651**
# Pull Request Summary

## Check List

- [x] deploying contract
  - [x] implement file upload module including dropzone
  - [x] fix legacy hooks
  - [x] Add loading, error message ui module
- [x] Add existing code hash
- [x] render contracts

## Change Log

This pull request makes the following changes:

### Adds

- deploying contract
- implement file upload module including dropzone
- Add loading, error message ui module
- Add existing code hash
- render contracts

### Fixes

- fix legacy hooks
- resolve contract-util issue. https://github.com/polkadot-js/api/issues/3591

### Changes

https://user-images.githubusercontent.com/1606820/120152498-86253d80-c228-11eb-97bd-9b2d61253c6b.mp4

for guidance

### CODE REVIEW

**Comment by sirius651** [comment]
while linting, it seems that the version of polkadot@api(v4.12.1) and node.js(v12) is not compatible.

File: src/components/dapps/CreateDapps.vue [hunk 1]
```diff
@@ -38,28 +38,35 @@
...
+    :address="defaultAccount"
+  />
+  <ModalCodeHash
+    v-if="modalCodeHash"
+    v-model:isOpen="modalCodeHash"
+    :address="defaultAccount"
+  />
 </template>
 
 <script lang="ts">
```
Comment (hoonsubin) [inline src/components/dapps/CreateDapps.vue:60]: The `CreateDapps.vue` component should be named `CreateDappsTab.vue` to prevent confusion. This applies to all other Tab components as well.

File: src/components/dapps/InputFile.vue [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        };
+
+        reader.readAsArrayBuffer(file);
+      });
+    };
+
+    const { getRootProps, getInputProps, ...rest } = useDropzone({
+      onDrop,
+      multiple: false,
+      accept: '.contract',
```
Comment (hoonsubin) [inline src/components/dapps/InputFile.vue]: ink! contracts are made out of `.json` abi file and `.wasm` binary file. `.contract` file is not used.
https://polkadot.js.org/docs/api-contract/start/code
I wrote this on my next review, but I think it's best to make the file extension a prop of the component.
You can reference this line of code in polkadot/api
https://github.com/polkadot-js/api/blob/92bde9de6f65d06dd9bb593627787f7fb2b390ac/packages/api-contract/src/base/Code.ts#L39

File: src/hooks/params/initValue.ts [hunk 3]
```diff
@@ -0,0 +1,159 @@
+import type { Registry, TypeDef } from '@polkadot/types/types';
+
+import { getTypeDef } from '@polkadot/types';
+import { TypeDefInfo } from '@polkadot/types/types';
+import { BN_ZERO, isBn } from '@polkadot/util';
+
+const warnList: string[] = [];
+
+export default function getInitValue(
```
Comment (hoonsubin) [inline src/hooks/params/initValue.ts]: The `polkadot/api` already does chain type parsing when it is initiated https://github.com/polkadot-js/api/blob/3273556f7baefa2301d0a2aa5c47e39d50598ad6/packages/api/src/base/Init.ts#L82.

I see a lot of repeating code and boilerplating that is already handled by the API library. I suggest using `OverideBundleDefinition` https://github.com/polkadot-js/api/blob/3c9b7de6c7fa214207a914397cdc03ceb96a6695/packages/types/src/types/registry.ts#L52 if that is what you are looking for.

For contracts, instantiating the `Code` class with the abi and wasm binary will automatically create the types for you https://github.com/polkadot-js/api/blob/a9c9fb5769dec7ada8612d6068cf69de04aa15ed/packages/api-contract/src/promise/Code.ts#L13

File: src/hooks/params/values.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+): RawParam {
+  const value = getInitValue(registry, param.type);
+
+  return {
+    isValid: !isUndefined(value),
+    value,
+  };
+}
+
+export default function createValues(
```
Comment (hoonsubin) [inline src/hooks/params/values.ts]: ~I know this is used in `ModalCreateDapps.vue` right now for parsing the abi params from the user. But this is all handled by `polkadot/api`. So this entire script shouldn't be needed, or convert it to a helper script for handling contract code generation.~

The function name `createValue` is too abstract and hard to understand what it does. Please don't use it if we have a function called `createValues` already exported.

File: src/hooks/signer/usePendingTxDeprecated.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    console.error(error);
+
+    return {
+      error: error as Error,
+      status: 'error',
+    };
+  }
+}
+
+async function sendRpc(
```
Comment (hoonsubin) [inline src/hooks/signer/usePendingTxDeprecated.ts]: the function name `submitRpc` and `sendRpc` is too close and confusing.

File: src/hooks/signer/useSendTx.ts [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+import { keyring } from '@polkadot/ui-keyring';
+import { web3FromSource } from '@polkadot/extension-dapp';
+import AccountSigner from './AccountSigner';
+import { useApi } from '@/hooks';
+import BN from 'bn.js';
+import { cacheUnlock, extractExternal, handleTxResults } from './util';
+
+export default function useSendTx() {
+  async function submitRpc(
+    api: ApiPromise,
```
Comment (hoonsubin) [inline src/hooks/signer/useSendTx.ts]: This applies to all the other hooks functions inside the `hooks` holder with the prefix `use` in their name, but if you want to create a hook function, please utilize the Context and Provider pattern with Vue3 so that we don't have to pass `api: ApiPromise` as a parameter (this will result in deep prop-dirlling).

We have a `polkadotContext.ts` function https://github.com/PlasmNetwork/plasm-apps-vue/blob/9d0538083fe77423370fdec5351de34e010047c4/src/api/polkadot/polkadotContext.ts#L11 (may ned improvements though), so please utilize this. `import { useApi } from '@/hooks';` It helps the project scale with additional pages and components.
I'll link to some resources that might help you.

- https://hackernoon.com/context-and-provider-pattern-using-vue-3-composition-api-c67m3vgx
- https://v3.vuejs.org/guide/component-provide-inject.html
- https://www.codemag.com/Article/2101091/The-Complete-Guide-to-Provide-Inject-API-in-Vue-3-Part-1

In short, we don't want to pass the entire API instance for using hooks.

File: src/hooks/useAbi.ts [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  registry?: {
+    strings?: any;
+  };
+}
+
+export default function useAbi(source: Code | null = null, isRequired = false) {
+  const { api } = useApi();
+
+  const registry = api?.value?.registry;
+  const chainProperties = registry?.getChainProperties() as
```
Comment (hoonsubin) [inline src/hooks/useAbi.ts:50]: I think it's better to abstract the chain registry data as its own hook (ex: `useRegistry`), since this data should not change that often, and I can imagine a lot of boilerplating going on for other methods that reads chain metadata.

File: src/components/dapps/InputFile.vue [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+import { hexToU8a, isHex, u8aToString } from '@polkadot/util';
+import { useDropzone } from 'vue3-dropzone';
+
+export default defineComponent({
+  components: {
+    IconDocument,
+  },
+  props: {
+    file: {
+      type: Object,
```
Comment (hoonsubin) [inline src/components/dapps/InputFile.vue:37]: I recommend making the props as a string array of uploadable file extensions so we can make it reusable.

File: src/components/dapps/ModalCreateDapps.vue [hunk 9]
```diff
@@ -223,41 +270,16 @@
...
                   </div>
-                </div>
+                </div> -->
 
                 <div>
                   <label
                     class="block text-sm font-medium text-gray-500 dark:text-darkGray-400 mb-2"
                   >
-                    Json file
+                    Contract file
```
Comment (hoonsubin) [inline src/components/dapps/ModalCreateDapps.vue:279]: When creating a new smart contract, the user should upload the abi json file and the wasm binary file. But I only see one upload panel. Where can the user upload their wasm file?

File: src/components/dapps/ModalCreateDapps.vue [hunk 10]
```diff
@@ -284,32 +306,371 @@
...
+        });
+        return;
+      }
+
+      const abiData = abi.value as Abi | AnyJson;
+
+      console.log('s', abiData);
+      console.log('w', wasm.value);
+
+      const code = new CodePromise(api?.value, abiData, wasm.value);
```
Comment (hoonsubin) [inline src/components/dapps/ModalCreateDapps.vue]: You are only making a local instance of a smart contract by doing this
```ts
const code = new CodePromise(api, abi, wasm);
```

To deploy the contract to the chain, you must sign and send the blueprint

```ts
// Deploy the WASM, retrieve a Blueprint
let blueprint;

// createBlueprint is a normal submittable, so use signAndSend
// with an known Alice keypair (as per the API samples)
const unsub = await code
  .createBlueprint()
  .signAndSend(alicePair, (result) => {
    if (result.status.isInBlock || result.status.isFinalized) {
      // here we have an additional field in the result, containing the blueprint
      blueprint = result.blueprint;
      unsub();
    }
  })
```
reference: https://polkadot.js.org/docs/api-contract/start/code

File: src/components/dapps/ModalCreateDapps.vue [hunk 11]
```diff
@@ -284,32 +306,371 @@
...
+          []
+        );
+        const arrValues: any = pvalues?.map(({ value }) => value);
+        console.log('v', arrValues);
+
+        uploadTx =
+          code &&
+          abi.value?.constructors[constructorIndex].method &&
+          formData.endowment
+            ? code.tx[abi.value?.constructors[constructorIndex].method](
```
Comment (hoonsubin) [inline src/components/dapps/ModalCreateDapps.vue:502]: It may not be good to parse the contract methods this way. I recommend using a structure like this https://github.com/polkadot-js/apps/blob/0ece312038b78d0191a1747be78afb9342a64e17/packages/page-contracts/src/Codes/Upload.tsx#L185-L197

In short, to create the contract constructor parameter, you need `abi.constructors[constructorIndex].args` and `abi.registry`

File: src/components/dapps/ModalCodeHash.vue [hunk 12]
```diff
@@ -82,16 +82,64 @@
...
+      required: true,
+    },
+  },
   setup(props, { emit }) {
     const closeModal = () => {
       emit('update:is-open', false);
     };
 
+    const store = useStore();
+    const api = computed(() => store.getters.api);
```
Comment (hoonsubin) [inline src/components/dapps/ModalCodeHash.vue]: I noticed that the api instance from the global store and the API instance from the injected context is used throughout the codebase. Please remove the API state in the global store and change all existing references to use the `useApi` method instead. This is hard to debug.

https://github.com/PlasmNetwork/plasm-apps-vue/blob/9d0538083fe77423370fdec5351de34e010047c4/src/hooks/useApi.ts#L6

https://github.com/PlasmNetwork/plasm-apps-vue/blob/732acb169c0ca8d4d39a4b05d74d32bd5476c637/src/store/state.ts#L21

**Review by hoonsubin — CHANGES_REQUESTED** [review]
I might have missed some errors, but these are some of the important ones that I was able to find. Please fix them

File: src/components/dapps/ModalCreateDapps.vue (hunk 10 above)
Comment (sirius651) [reply src/components/dapps/ModalCreateDapps.vue]: I already tried this, but it seems `createBlueprint` method is deprecated. so, that's why I should had changed my mind to follow `canvas-ui` or `polkadot-explorer` codebase. they are not using this method anywhere.

Comment (hoonsubin) [reply src/components/dapps/ModalCreateDapps.vue]: @sirius651 You're right. Sorry for my misinformation!

The smart contract instantiation on chain looks different.
I can see that from this code https://github.com/paritytech/substrate/blob/6aaa03130da8abbe7c0aa5cf4c3f10557713f9d6/frame/contracts/src/lib.rs#L329-L336
you have to send the contract code, endowment, constructor parameter to deploy it to the chain.

The `createBluePrintTx()` has been simplified and now is part of the constructor tx call of the contract.
https://github.com/polkadot-js/api/blob/92bde9de6f65d06dd9bb593627787f7fb2b390ac/packages/api-contract/src/base/Code.ts#L50

So your method of using `CodePromise.tx.<contractor>.method()` is correct.
https://github.com/polkadot-js/apps/blob/master/packages/page-contracts/src/Codes/Upload.tsx#L90

But I still can't seem to deploy my contracts... I'll try to look into this again!
Having said that, it would be great if you can organize the code structure by separating the contract code-related functions as a `useHook` function or a separate component like this https://github.com/polkadot-js/apps/tree/master/packages/page-contracts/src/Codes. The code is getting too long to debug...

File: src/components/dapps/InputFile.vue (hunk 2 above)
Comment (sirius651) [reply src/components/dapps/InputFile.vue]: but, `canvas-ui` ( https://paritytech.github.io/canvas-ui/#/instantiate/new ) or `polkadot-explorer` ( https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A9944#/contracts ) is not support `.wasm` file. but they support `.contract` or `metadata.json` abi file.

File: src/hooks/signer/useSendTx.ts (hunk 6 above)
Comment (sirius651) [reply src/hooks/signer/useSendTx.ts]: yeah, that makes sense. but the reason why I pass `api: ApiPromise` instead of using useApi is that `api` goes undefined when I using useApi func sometimes. it seems not working well. maybe should refactor all codes regarding `polkadotContext` related files.

File: src/hooks/signer/usePendingTxDeprecated.ts (hunk 5 above)
Comment (sirius651) [reply src/hooks/signer/usePendingTxDeprecated.ts]: removed this file

File: src/components/dapps/ModalCodeHash.vue (hunk 12 above)
Comment (sirius651) [reply src/components/dapps/ModalCodeHash.vue]: it's similar reason what I chose. https://github.com/PlasmNetwork/plasm-apps-vue/pull/24#discussion_r643277348 
if `useApi` is fixed, it would be resolved. will fix at first then

File: src/api/polkadot/connectApi.ts [hunk 13]
```diff
@@ -1,8 +1,7 @@
-import { ApiPromise } from '@polkadot/api';
+import { ApiPromise, WsProvider } from '@polkadot/api';
 import { keyring } from '@polkadot/ui-keyring';
 import { web3Accounts, web3Enable } from '@polkadot/extension-dapp';
 import { isTestChain } from '@polkadot/util';
-import { WsProvider } from '@polkadot/rpc-provider';
 import * as plasmDefinitions from '@plasm/types/interfaces/definitions';
```
Comment (hoonsubin) [inline src/api/polkadot/connectApi.ts]: You can fix the `instantiateWithCode(endowment, gas_limit, code, data, salt)` error by updating the `@plasm/types` to `^1.5.4-2`. Here is how you can use the new import

```ts

import chainTypeDefs from '@plasm/types';
...

export const connectApi = async (endpoint: string) => {
  const provider = new WsProvider(endpoint);
  const api = new ApiPromise({
    provider,
    types: {
      ...chainTypeDefs.dustyDefinitions,
      LookupSource: 'MultiAddress',
    },
  });

  try {
    await api.isReadyOrError;
  } catch (err) {
    console.error('err', err);
  }

  try {
    await loadAccounts(api);
  } catch (err) {
    console.error(err);
  }

  // load the web3 extension
  injectedPromise
    .then((): void => {})
    .catch((error: Error) => console.error(error));

  return api;
};
```

File: src/hooks/params/values.ts (hunk 4 above)
Comment (sirius651) [reply src/hooks/params/values.ts]: I'm not sure where can I handle on `polkadot/api` . but refactored this.

File: src/hooks/params/initValue.ts (hunk 3 above)
Comment (sirius651) [reply src/hooks/params/initValue.ts]: where can I use `OverideBundleDefinition` on it?

File: src/hooks/useAbi.ts (hunk 7 above)
Comment (sirius651) [reply src/hooks/useAbi.ts:50]: After investigating `registry` things, it doesn't seem necessary to separate it now. there're `Registry` and `MetaRegistry` in there. and, we can get access registry easily on `api` or `abi` instance.

File: src/hooks/params/initValue.ts (hunk 3 above)
Comment (hoonsubin) [reply src/hooks/params/initValue.ts]: Sorry, my comment was based on the outdated docs. I think the new method is different

File: src/hooks/params/values.ts (hunk 4 above)
Comment (hoonsubin) [reply src/hooks/params/values.ts]: Sorry for the previous comment. Now that I think about it, it is good to have a dynamic value parser not only for type registration, but for UI element generation as well. But I don't think this is a function that should be in `hooks/` folder? What about making this as a helper function?

The name `value` is too broad

**Review by hoonsubin — CHANGES_REQUESTED** [review]

File: src/hooks/params/values.ts (hunk 4 above)
Comment (sirius651) [reply src/hooks/params/values.ts]: yeah, it resolved

**Comment by sirius651** [comment]
@hoonsubin revolved all cases based on above feedbacks. maybe you can review it again. but, there's CI limiting issue I've got. 
```
warning Lockfile has incorrect entry for "@polkadot/api@^4.12.2". Ignoring it.
9
error Couldn't find any versions for "@polkadot/api" that matches "^4.12.2"
10
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

**Comment by hoonsubin** [comment]
> @hoonsubin revolved all cases based on above feedbacks. maybe you can review it again. but, there's CI limiting issue I've got.
> 
> ```
> warning Lockfile has incorrect entry for "@polkadot/api@^4.12.2". Ignoring it.
> 9
> error Couldn't find any versions for "@polkadot/api" that matches "^4.12.2"
> 10
> info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
> ```

The CI error is due to the dependency resolution with `@plasm/types` package. I'll fix it!

File: src/components/dapps/ModalCreateDapps.vue [hunk 14]
```diff
@@ -284,32 +310,310 @@
...
+        // );
+        // const weight = new BN(200000000000);
+        ///
+        console.log('endowment', formData.endowment);
+        console.log('weight', formData.weight);
+
+        console.log('code', code);
+        console.log('method', code.tx);
+
+        const constructorIndex = 0;
```
Comment (hoonsubin) [inline src/components/dapps/ModalCreateDapps.vue:491]: The constructor index doesn't seem to be user-definable like in the apps portal?
https://github.com/polkadot-js/apps/blob/0ece312038b78d0191a1747be78afb9342a64e17/packages/page-contracts/src/Codes/Upload.tsx#L189
If you are planning on fixing this for the next PR, please let me know.

File: src/components/dapps/ModalCreateDapps.vue [hunk 15]
```diff
@@ -284,32 +310,310 @@
...
+
+      const _onStart = () => {
+        console.log('_onStart');
+        store.commit(MutationTypes.SET_LOADING, true);
+      };
+
+      const _onSuccess = (result: CodeSubmittableResult) => {
+        console.log('_onSuccess', result);
+
+        const codeHash = result.blueprint?.codeHash;
```
Comment (hoonsubin) [inline src/components/dapps/ModalCreateDapps.vue:540]: The contract hash is generated and stored locally, but it is not deployed to the chain.

Comment (sirius651) [reply src/components/dapps/ModalCreateDapps.vue:540]: how can I verify whether this contract is deployed to the chain? because currently, it seems work similar with polkadot explorer

Comment (hoonsubin) [reply src/components/dapps/ModalCreateDapps.vue:540]: I tried deploying the contract from plasm apps vue and access the address from Polkadot-js apps portal. It fails to load the contract. If you can access the contract from both applications, please let me know! Because it could be an issue with my local environment (perhaps related to chain types?)

Comment (hoonsubin) [reply src/components/dapps/ModalCreateDapps.vue:540]: What about adding a unit test for checking API operations? We can either use mock data or real async calls to the node.
I'm working on a GitHub action for loading a local development node as part of the workflow (still WIP). https://github.com/hoonsubin/build-substrate-action
Having a unit test can definitely help development

Comment (hoonsubin) [reply src/components/dapps/ModalCreateDapps.vue:540]: @sirius651 I want to fix my statement above. I was able to check that the contract does deploy to the blockchain. But (outside of the issue of not being able to enter the constructor params) I can't sign and deploy with the address that I choose from the dropdown. So I think that was the error that I saw.

(in the contract, it says the creator is Alice, but I tried to deploy with a different account)
Screen Shot 2021-06-08 at 2 18 07 PM
Screen Shot 2021-06-08 at 2 17 50 PM

Comment (sirius651) [reply src/components/dapps/ModalCreateDapps.vue:540]: > What about adding a unit test for checking API operations? We can either use mock data or real async calls to the node.
> I'm working on a GitHub action for loading a local development node as part of the workflow (still WIP). https://github.com/hoonsubin/build-substrate-action
> Having a unit test can definitely help development

yeah, I totally agree on that. I'll set up unit test when I have time

Comment (sirius651) [reply src/components/dapps/ModalCreateDapps.vue:540]: > @sirius651 I want to fix my statement above. I was able to check that the contract does deploy to the blockchain. But (outside of the issue of not being able to enter the constructor params) I can't sign and deploy with the address that I choose from the dropdown. So I think that was the error that I saw.
> 

I found out that is bug. will be fixed on new branch : feature/deploy-contract
